
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Differential gene expression analysis &#8212; Consensus Pathway Analysis using Google Cloud Infrastructure</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Gene set and pathway" href="Module03-Gene_Set_and_Pathway.html" />
    <link rel="prev" title="Gene ID mapping" href="Module01-Gene_Mapping.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/Cover.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Consensus Pathway Analysis using Google Cloud Infrastructure</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Overview
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting Started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="VertexAI.html">
   Setting up Google Cloud environment
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 01
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module01-GEO_Data_Processing.html">
   GEO data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Module01-Gene_Mapping.html">
   Gene ID mapping
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 02
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Differential gene expression analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 03
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module03-Gene_Set_and_Pathway.html">
   Gene set and pathway
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 04
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module04-Pathway_Analysis.html">
   Pathway Analysis
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/executablebooks/jupyter-book/master?urlpath=tree/docs/Module02-DE_Analysis.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/executablebooks/jupyter-book"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2FModule02-DE_Analysis.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/Module02-DE_Analysis.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#learning-objectives">
   Learning Objectives:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#grouping-samples-and-setting-up-the-design-matrix">
   Grouping samples and setting up the design matrix.
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#de-analysis-using-limma">
   DE analysis using limma.
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#de-analysis-using-t-test">
   DE analysis using t-test
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#de-analysis-using-edger">
   DE analysis using edgeR
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#visualization-of-of-differential-expression-results">
   Visualization of of differential expression results.
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Differential gene expression analysis</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#learning-objectives">
   Learning Objectives:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#grouping-samples-and-setting-up-the-design-matrix">
   Grouping samples and setting up the design matrix.
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#de-analysis-using-limma">
   DE analysis using limma.
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#de-analysis-using-t-test">
   DE analysis using t-test
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#de-analysis-using-edger">
   DE analysis using edgeR
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#visualization-of-of-differential-expression-results">
   Visualization of of differential expression results.
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="differential-gene-expression-analysis">
<h1>Differential gene expression analysis<a class="headerlink" href="#differential-gene-expression-analysis" title="Permalink to this headline">#</a></h1>
<p><img alt="" src="_images/DE_pipeline.jpg" /></p>
<p>The next step in the pathway analysis workflow is the differential expression (DE)  analysis. The goal of DE testing is to determine which genes are expressed at different levels between two or more biological conditions. These genes can offer biological insight into the processes affected by the condition(s) of interest. DE analysis means taking the normalised read count data and performing statistical analysis to discover quantitative changes in expression levels between experimental groups. For example, we use statistical testing to decide whether, for a given gene, an observed difference in read counts is significant, that is, whether it is greater than what would be expected just due to natural random variation. This learning module demonstrates a computational workflow for the detection of DE genes from RNA-Seq data.</p>
<section id="learning-objectives">
<h2>Learning Objectives:<a class="headerlink" href="#learning-objectives" title="Permalink to this headline">#</a></h2>
<ol class="simple">
<li><p>Assigning samples into groups and set up design matrix.</p></li>
<li><p>Performing differential expression (DE) analysis using <code class="docutils literal notranslate"><span class="pre">limma</span></code>, <code class="docutils literal notranslate"><span class="pre">t-test</span></code>, <code class="docutils literal notranslate"><span class="pre">edgeR</span></code> and <code class="docutils literal notranslate"><span class="pre">DEseq</span></code> R package.</p></li>
<li><p>Filtering and exporting the results table.</p></li>
<li><p>Further visualisation</p></li>
</ol>
</section>
<section id="grouping-samples-and-setting-up-the-design-matrix">
<h2>Grouping samples and setting up the design matrix.<a class="headerlink" href="#grouping-samples-and-setting-up-the-design-matrix" title="Permalink to this headline">#</a></h2>
<p>In this section we will perform differential expression (DF) analysis to analyze the <code class="docutils literal notranslate"><span class="pre">GSE48350</span></code> dataset that we have used in the previous chapter.
Recall from the <a class="reference internal" href="Module01-GEO_Data_Processing.html"><span class="doc std std-doc"><strong>GEO Data processing</strong> </span></a> section,<code class="docutils literal notranslate"><span class="pre">GSE48350</span></code> contains
sequenced data of Human Alzheimer’s Disease using microarray sequencing technology collected from 4 brain regions.
Here we will create two groups: <em>“condition”</em> and <em>“control”</em> from the Entorhinal Cortex region. First, we need to load necessary R libraries to download the dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">suppressMessages</span><span class="p">({</span><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
    <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
  <span class="nf">suppressWarnings</span><span class="p">(</span><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;GEOquery&quot;</span><span class="p">,</span> <span class="n">update</span> <span class="o">=</span> <span class="bp">T</span><span class="p">))</span>
  <span class="nf">suppressWarnings</span><span class="p">(</span><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;limma&quot;</span><span class="p">,</span> <span class="n">update</span> <span class="o">=</span> <span class="bp">T</span><span class="p">))</span>
<span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">suppressPackageStartupMessages</span><span class="p">({</span>
  <span class="nf">library</span><span class="p">(</span><span class="s">&quot;GEOquery&quot;</span><span class="p">)</span>
  <span class="nf">library</span><span class="p">(</span><span class="s">&quot;limma&quot;</span><span class="p">)</span>
<span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<p>Similar to the <a class="reference internal" href="Module01-GEO_Data_Processing.html"><span class="doc std std-doc"><strong>GEO Data processing</strong> </span></a> section, we can use <a href="https://bioconductor.org/packages/release/bioc/html/GEOquery.html">GEOquery</a> R package to download GEO data and select samples generated from Affymetrix Human Genome U133 Plus 2.0 Array platform.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load series and platform data from GEO</span>
<span class="n">gset</span> <span class="o">&lt;-</span> <span class="nf">suppressMessages</span><span class="p">(</span><span class="nf">getGEO</span><span class="p">(</span><span class="s">&quot;GSE48350&quot;</span><span class="p">,</span> <span class="n">GSEMatrix</span> <span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">AnnotGPL</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span>
<span class="c1"># Check how many platform available</span>
<span class="nf">if </span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">gset</span><span class="p">)</span> <span class="o">&gt;</span> <span class="m">1</span><span class="p">)</span> <span class="n">idx</span> <span class="o">&lt;-</span> <span class="nf">grep</span><span class="p">(</span><span class="s">&quot;GPL570&quot;</span><span class="p">,</span> <span class="nf">attr</span><span class="p">(</span><span class="n">gset</span><span class="p">,</span> <span class="s">&quot;names&quot;</span><span class="p">))</span> <span class="n">else</span> <span class="n">idx</span> <span class="o">&lt;-</span> <span class="m">1</span>
<span class="n">gset</span> <span class="o">&lt;-</span> <span class="n">gset</span><span class="p">[[</span><span class="n">idx</span><span class="p">]]</span>
<span class="c1"># Get expression data matric</span>
<span class="n">expression_data</span> <span class="o">&lt;-</span> <span class="nf">exprs</span><span class="p">(</span><span class="n">gset</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We can check for number of samples and genes using this command:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">print</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;#Genes: &quot;</span><span class="p">,</span><span class="nf">dim</span><span class="p">(</span><span class="n">expression_data</span><span class="p">)[</span><span class="m">1</span><span class="p">],</span><span class="s">&quot; - #Samples: &quot;</span><span class="p">,</span><span class="nf">dim</span><span class="p">(</span><span class="n">expression_data</span><span class="p">)[</span><span class="m">2</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;#Genes: 54675 - #Samples: 253&quot;
</pre></div>
</div>
</div>
</div>
<p>Now we need to select the samples that belong to entorhinal cortex region using following command</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get samples information</span>
<span class="n">samples</span> <span class="o">&lt;-</span> <span class="nf">pData</span><span class="p">(</span><span class="n">gset</span><span class="p">)</span>
<span class="c1"># Select sample from Entorhinal Cortex region</span>
<span class="n">idx</span> <span class="o">&lt;-</span> <span class="nf">grep</span><span class="p">(</span><span class="s">&quot;entorhinal&quot;</span><span class="p">,</span> <span class="n">samples</span><span class="o">$</span><span class="n">`brain region:ch1`</span><span class="p">)</span>
<span class="n">gset</span> <span class="o">=</span> <span class="n">gset</span><span class="p">[</span> <span class="p">,</span><span class="n">idx</span><span class="p">]</span>
<span class="n">samples</span> <span class="o">=</span> <span class="n">samples</span><span class="p">[</span><span class="n">idx</span><span class="p">,]</span>
<span class="c1"># Get expression data for samples collected from entorhinal cortex region</span>
<span class="n">expression_data</span> <span class="o">=</span> <span class="n">expression_data</span><span class="p">[,</span><span class="n">idx</span><span class="p">]</span>
<span class="c1"># Print out new number genes and samples</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;#Genes: &quot;</span><span class="p">,</span><span class="nf">dim</span><span class="p">(</span><span class="n">expression_data</span><span class="p">)[</span><span class="m">1</span><span class="p">],</span><span class="s">&quot; - #Samples: &quot;</span><span class="p">,</span><span class="nf">dim</span><span class="p">(</span><span class="n">expression_data</span><span class="p">)[</span><span class="m">2</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;#Genes: 54675 - #Samples: 54&quot;
</pre></div>
</div>
</div>
</div>
<p>We can see that there are 54 samples that were collected from entorhinal cortex region. Next, we need to perform a <span class="math notranslate nohighlight">\(log_2\)</span> transformation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate quantile values from the data</span>
<span class="n">qx</span> <span class="o">&lt;-</span> <span class="nf">as.numeric</span><span class="p">(</span><span class="nf">quantile</span><span class="p">(</span><span class="n">expression_data</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="n">.</span><span class="p">,</span> <span class="m">0.25</span><span class="p">,</span> <span class="m">0.5</span><span class="p">,</span> <span class="m">0.75</span><span class="p">,</span> <span class="m">0.99</span><span class="p">,</span> <span class="m">1.0</span><span class="p">),</span> <span class="n">na.rm</span><span class="o">=</span><span class="bp">T</span><span class="p">))</span>
<span class="c1"># Perform log2 transformation</span>
<span class="n">LogC</span> <span class="o">&lt;-</span> <span class="p">(</span><span class="n">qx</span><span class="p">[</span><span class="m">5</span><span class="p">]</span> <span class="o">&gt;</span> <span class="m">100</span><span class="p">)</span> <span class="o">||</span> <span class="p">(</span><span class="n">qx</span><span class="p">[</span><span class="m">6</span><span class="p">]</span><span class="o">-</span><span class="n">qx</span><span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="m">50</span> <span class="o">&amp;&amp;</span> <span class="n">qx</span><span class="p">[</span><span class="m">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">)</span>
<span class="nf">if </span><span class="p">(</span><span class="n">LogC</span><span class="p">)</span> <span class="p">{</span> <span class="n">expression_data</span><span class="p">[</span><span class="nf">which</span><span class="p">(</span><span class="n">expression_data</span> <span class="o">&lt;=</span> <span class="m">0</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="kc">NaN</span>
<span class="nf">exprs</span><span class="p">(</span><span class="n">gset</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">log2</span><span class="p">(</span><span class="n">expression_data</span><span class="p">)}</span>
</pre></div>
</div>
</div>
</div>
<p>In order to perform DE analysis, we need to group patients into two groups: <em>“condition”</em> and <em>“control”</em>. From the samples source name, patients diagnosed with Alzheimer’s Disease are annotated with ‘AD’. Therefore, any samples that contain string ‘AD’ are labelled by ‘d’ (condition/disease) and the remaining samples are assigned to ‘c’ (control/normal). The code to group the samples is presented below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select diease samples</span>
<span class="n">disease_idx</span> <span class="o">&lt;-</span> <span class="nf">grep</span><span class="p">(</span><span class="s">&quot;AD&quot;</span><span class="p">,</span> <span class="n">samples</span><span class="o">$</span><span class="n">source_name_ch1</span><span class="p">)</span>
<span class="c1"># Create a vector to store label</span>
<span class="n">groups</span> <span class="o">&lt;-</span> <span class="nf">rep</span><span class="p">(</span><span class="s">&quot;X&quot;</span><span class="p">,</span> <span class="nf">nrow</span><span class="p">(</span><span class="n">samples</span><span class="p">))</span>
<span class="c1"># Annotate a diasease sameples as &quot;d&quot;</span>
<span class="n">groups</span><span class="p">[</span><span class="n">disease_idx</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="s">&quot;d&quot;</span>
<span class="c1"># Control samples are labeled as &quot;c&quot;</span>
<span class="n">groups</span><span class="p">[</span><span class="nf">which</span><span class="p">(</span><span class="n">groups</span><span class="o">!=</span><span class="s">&quot;d&quot;</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="s">&quot;c&quot;</span>
<span class="n">groups</span> <span class="o">&lt;-</span> <span class="nf">factor</span><span class="p">(</span><span class="n">groups</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="de-analysis-using-limma">
<h2>DE analysis using limma.<a class="headerlink" href="#de-analysis-using-limma" title="Permalink to this headline">#</a></h2>
<p>Now, we need to specify the model to be fitted using <code class="docutils literal notranslate"><span class="pre">limma</span></code>. Linear modelling in <code class="docutils literal notranslate"><span class="pre">limma</span></code> is carried out using the <code class="docutils literal notranslate"><span class="pre">lmFit</span></code> and <code class="docutils literal notranslate"><span class="pre">contrasts.fit</span></code> functions originally written for application to microarrays. The functions can be used for both microarray and RNA-seq data and fit a separate model to the expression values for each gene. Next, empirical Bayes moderation is carried out using <code class="docutils literal notranslate"><span class="pre">eBayes</span></code> function.  Empirical Bayes moderation borrows information across all the genes to obtain more precise estimates of gene-wise variability. The model’s residual variances are plotted against average expression values in the next figure. It can be seen from this plot that the variance is no longer dependent on the mean expression level.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Add groups to gset object</span>
<span class="n">gset</span><span class="o">$</span><span class="n">group</span> <span class="o">&lt;-</span> <span class="n">groups</span>
<span class="n">design</span> <span class="o">&lt;-</span> <span class="nf">model.matrix</span><span class="p">(</span><span class="o">~</span><span class="n">group</span> <span class="o">+</span> <span class="m">0</span><span class="p">,</span> <span class="n">gset</span><span class="p">)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">design</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">levels</span><span class="p">(</span><span class="n">groups</span><span class="p">)</span>
<span class="n">fit</span> <span class="o">&lt;-</span> <span class="nf">lmFit</span><span class="p">(</span><span class="n">gset</span><span class="p">,</span> <span class="n">design</span><span class="p">)</span>  <span class="c1"># fit linear model</span>
<span class="c1"># set up contrasts of interest and recalculate model coefficients</span>
<span class="n">cts</span> <span class="o">&lt;-</span> <span class="nf">paste</span><span class="p">(</span><span class="s">&quot;c&quot;</span><span class="p">,</span> <span class="s">&quot;d&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;-&quot;</span><span class="p">)</span>
<span class="n">cont.matrix</span> <span class="o">&lt;-</span> <span class="nf">makeContrasts</span><span class="p">(</span><span class="n">contrasts</span><span class="o">=</span><span class="n">cts</span><span class="p">,</span> <span class="n">levels</span><span class="o">=</span><span class="n">design</span><span class="p">)</span>
<span class="n">fit2</span> <span class="o">&lt;-</span> <span class="nf">contrasts.fit</span><span class="p">(</span><span class="n">fit</span><span class="p">,</span> <span class="n">cont.matrix</span><span class="p">)</span>

<span class="c1"># compute statistics and table of top significant genes</span>
<span class="n">fit2</span> <span class="o">&lt;-</span> <span class="nf">eBayes</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="m">0.01</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>To see the number of significantly up- and down-regulated genes, we can use the code below to generate the summary table.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">dt</span> <span class="o">&lt;-</span> <span class="nf">decideTests</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span><span class="n">p.value</span><span class="o">=</span><span class="m">0.05</span><span class="p">)</span>
<span class="nf">summary</span><span class="p">(</span><span class="n">dt</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>         c-d
Down     293
NotSig 53893
Up       489
</pre></div>
</div>
</div>
</div>
<p>Here, significance is defined using an adjusted p-value cutoff that is set at 5% by default. For the comparison between expression levels in <em>“condition”</em> (d) and <em>“control”</em> (c), 293 genes are found to be down-regulated and 489 genes are up-regulated. We also can extract a table of the top-ranked genes from a linear model fit using <code class="docutils literal notranslate"><span class="pre">topTable</span></code> function. By default, <code class="docutils literal notranslate"><span class="pre">topTable</span></code> arranges genes from smallest to largest adjusted p-value with associated gene information, log-FC, average log-CPM, moderated t-statistic, raw and adjusted p-value for each gene. The number of top genes displayed can be specified, where <code class="docutils literal notranslate"><span class="pre">number=Inf</span></code> includes all genes. In the following code, we will select top 1,000 genes are differentially expressed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">top_genes</span> <span class="o">&lt;-</span> <span class="nf">topTable</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">adjust</span><span class="o">=</span><span class="s">&quot;fdr&quot;</span><span class="p">,</span> <span class="n">sort.by</span><span class="o">=</span><span class="s">&quot;B&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="m">1000</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>By default, the <code class="docutils literal notranslate"><span class="pre">topTable</span></code> function generates a dataframe that consists of DE genes, statistic and other features. We do not need all the features to perform pathway and enrichment analyses in the later module. Therefore, we will select a subset of <code class="docutils literal notranslate"><span class="pre">topTable</span></code> output by using the following command:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">top_genes</span> <span class="o">&lt;-</span> <span class="nf">subset</span><span class="p">(</span><span class="n">top_genes</span><span class="p">,</span> <span class="n">select</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="p">,</span><span class="s">&quot;Gene.symbol&quot;</span><span class="p">,</span><span class="s">&quot;adj.P.Val&quot;</span><span class="p">,</span><span class="s">&quot;P.Value&quot;</span><span class="p">,</span><span class="s">&quot;t&quot;</span><span class="p">,</span><span class="s">&quot;B&quot;</span><span class="p">,</span><span class="s">&quot;logFC&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<p>In order to perform pathway and enrichment analyses, we only focus on the genes that are statistically significant and the probe IDs that have gene symbols. We can select those genes using the following command:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">top_genes</span> <span class="o">&lt;-</span> <span class="n">top_genes</span><span class="p">[</span><span class="nf">which</span><span class="p">(</span><span class="n">top_genes</span><span class="o">$</span><span class="n">adj.P.Val</span><span class="o">&lt;</span><span class="m">0.05</span><span class="p">),]</span>
<span class="n">top_genes</span> <span class="o">&lt;-</span> <span class="n">top_genes</span><span class="p">[</span><span class="nf">which</span><span class="p">(</span><span class="n">top_genes</span><span class="o">$</span><span class="n">Gene.symbol</span><span class="o">!=</span><span class="s">&quot;&quot;</span><span class="p">),]</span>
<span class="nf">head</span><span class="p">(</span><span class="n">top_genes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 6 × 7</caption>
<thead>
	<tr><th></th><th scope=col>ID</th><th scope=col>Gene.symbol</th><th scope=col>adj.P.Val</th><th scope=col>P.Value</th><th scope=col>t</th><th scope=col>B</th><th scope=col>logFC</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>224687_at</th><td>224687_at  </td><td>ANKIB1   </td><td>1.575045e-38</td><td>5.761483e-43</td><td>-42.51195</td><td>71.37488</td><td>-5.0270459</td></tr>
	<tr><th scope=row>234109_x_at</th><td>234109_x_at</td><td>ONECUT3  </td><td>3.581024e-28</td><td>3.929793e-32</td><td>-26.08643</td><td>56.26670</td><td>-0.5401491</td></tr>
	<tr><th scope=row>230832_at</th><td>230832_at  </td><td>RTF1     </td><td>5.202052e-28</td><td>6.660149e-32</td><td> 25.81105</td><td>55.88888</td><td> 1.5886452</td></tr>
	<tr><th scope=row>236977_at</th><td>236977_at  </td><td>LOC646588</td><td>3.302438e-25</td><td>5.436112e-29</td><td> 22.51884</td><td>50.90379</td><td> 1.2085885</td></tr>
	<tr><th scope=row>211318_s_at</th><td>211318_s_at</td><td>RAE1     </td><td>6.039523e-25</td><td>1.104622e-28</td><td> 22.19198</td><td>50.35766</td><td> 2.9567118</td></tr>
	<tr><th scope=row>219554_at</th><td>219554_at  </td><td>RHCG     </td><td>2.166963e-24</td><td>4.359689e-28</td><td> 21.56994</td><td>49.29043</td><td> 0.9245753</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>Now, we can use the following script to save the DE genes list to use in the later modules.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">saveRDS</span><span class="p">(</span><span class="n">top_genes</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s">&quot;./data/DE_genes.rds&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="de-analysis-using-t-test">
<h2>DE analysis using t-test<a class="headerlink" href="#de-analysis-using-t-test" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">suppressMessages</span><span class="p">({</span>
  <span class="nf">suppressWarnings</span><span class="p">(</span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;matrixTests&quot;</span><span class="p">,</span><span class="n">quiet</span><span class="o">=</span> <span class="bp">T</span><span class="p">))</span>
<span class="p">})</span>
<span class="c1"># Load matrixTests package</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">({</span><span class="nf">library</span><span class="p">(</span><span class="s">&quot;matrixTests&quot;</span><span class="p">)})</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>package &#39;matrixTests&#39; successfully unpacked and MD5 sums checked
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">expression_data</span>
<span class="n">X</span> <span class="o">&lt;-</span> <span class="n">count</span><span class="p">[,</span><span class="n">groups</span><span class="o">==</span><span class="s">&quot;c&quot;</span><span class="p">]</span>
<span class="n">Y</span> <span class="o">&lt;-</span> <span class="n">count</span><span class="p">[,</span><span class="n">groups</span><span class="o">==</span><span class="s">&quot;d&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">res</span> <span class="o">&lt;-</span> <span class="nf">row_t_equalvar</span><span class="p">(</span><span class="n">X</span><span class="p">,</span><span class="n">Y</span><span class="p">,</span><span class="n">alternative</span> <span class="o">=</span> <span class="s">&quot;two.sided&quot;</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="m">0</span><span class="p">,</span> <span class="n">conf.level</span> <span class="o">=</span> <span class="m">0.95</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">res</span> <span class="o">&lt;-</span> <span class="n">res</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">res</span><span class="o">$</span><span class="n">pvalue</span><span class="p">),]</span>
<span class="n">res</span> <span class="o">&lt;-</span> <span class="n">res</span><span class="p">[</span><span class="n">res</span><span class="o">$</span><span class="n">pvalue</span><span class="o">&lt;</span><span class="m">0.05</span><span class="p">,]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="de-analysis-using-edger">
<h2>DE analysis using edgeR<a class="headerlink" href="#de-analysis-using-edger" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">suppressMessages</span><span class="p">({</span> <span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
  <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
  <span class="nf">suppressWarnings</span><span class="p">(</span><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;edgeR&quot;</span><span class="p">,</span> <span class="n">update</span> <span class="o">=</span> <span class="bp">F</span><span class="p">))</span>
<span class="p">})</span>

<span class="c1"># Load edgeR package</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">({</span><span class="nf">library</span><span class="p">(</span><span class="s">&quot;edgeR&quot;</span><span class="p">)})</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">expression_data</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">dge</span> <span class="o">&lt;-</span> <span class="nf">DGEList</span><span class="p">(</span><span class="n">counts</span> <span class="o">=</span> <span class="n">count</span><span class="p">,</span> <span class="n">group</span> <span class="o">=</span> <span class="nf">factor</span><span class="p">(</span><span class="n">gset</span><span class="o">$</span><span class="n">group</span><span class="p">))</span>
<span class="n">keep</span> <span class="o">&lt;-</span> <span class="nf">filterByExpr</span><span class="p">(</span><span class="n">y</span> <span class="o">=</span> <span class="n">dge</span><span class="p">)</span>
<span class="n">dge</span> <span class="o">&lt;-</span> <span class="n">dge</span><span class="p">[</span><span class="n">keep</span><span class="p">,</span><span class="n">keep.lib.sizes</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">]</span>
<span class="n">dge</span> <span class="o">&lt;-</span> <span class="nf">calcNormFactors</span><span class="p">(</span><span class="n">object</span> <span class="o">=</span> <span class="n">dge</span><span class="p">)</span>
<span class="n">dge</span> <span class="o">&lt;-</span> <span class="nf">estimateDisp</span><span class="p">(</span><span class="n">y</span> <span class="o">=</span> <span class="n">dge</span><span class="p">)</span>
<span class="n">et</span> <span class="o">&lt;-</span> <span class="nf">exactTest</span><span class="p">(</span><span class="n">object</span> <span class="o">=</span> <span class="n">dge</span><span class="p">)</span>
<span class="n">top_degs</span> <span class="o">=</span> <span class="nf">topTags</span><span class="p">(</span><span class="n">object</span> <span class="o">=</span> <span class="n">et</span><span class="p">,</span> <span class="n">n</span> <span class="o">=</span> <span class="s">&quot;Inf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Using classic mode.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">write.csv</span><span class="p">(</span><span class="n">et</span><span class="o">$</span><span class="n">table</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s">&quot;./data/edgeR_Results.csv&quot;</span><span class="p">)</span>
<span class="nf">saveRDS</span><span class="p">(</span><span class="n">et</span><span class="o">$</span><span class="n">table</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s">&quot;./data/edgeR_Results.rds&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">summary</span><span class="p">(</span><span class="nf">decideTests</span><span class="p">(</span><span class="n">object</span> <span class="o">=</span> <span class="n">et</span><span class="p">,</span> <span class="n">lfc</span> <span class="o">=</span> <span class="m">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>       d-c
Down     0
NotSig   5
Up       3
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">## DE analysis using DESeq2</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">suppressMessages</span><span class="p">({</span> <span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
  <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
  <span class="nf">suppressWarnings</span><span class="p">(</span><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;DESeq2&quot;</span><span class="p">,</span> <span class="n">update</span> <span class="o">=</span> <span class="bp">F</span><span class="p">))</span>
<span class="p">})</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">({</span><span class="nf">library</span><span class="p">(</span><span class="s">&quot;DESeq2&quot;</span><span class="p">)})</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">coldata</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span>
  <span class="n">sample</span> <span class="o">=</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">expression_data</span><span class="p">),</span>
  <span class="n">condition</span> <span class="o">=</span> <span class="nf">as.factor</span><span class="p">(</span><span class="n">groups</span><span class="p">),</span>
  <span class="n">row.names</span> <span class="o">=</span> <span class="s">&quot;sample&quot;</span> <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">head</span><span class="p">(</span><span class="n">coldata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 6 × 1</caption>
<thead>
	<tr><th></th><th scope=col>condition</th></tr>
	<tr><th></th><th scope=col>&lt;fct&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>GSM300173</th><td>c</td></tr>
	<tr><th scope=row>GSM300177</th><td>c</td></tr>
	<tr><th scope=row>GSM300181</th><td>c</td></tr>
	<tr><th scope=row>GSM300186</th><td>c</td></tr>
	<tr><th scope=row>GSM300189</th><td>c</td></tr>
	<tr><th scope=row>GSM300192</th><td>c</td></tr>
</tbody>
</table>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">range</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>0.01</li><li>1054.465</li></ol>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">dds</span> <span class="o">&lt;-</span> <span class="nf">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">countData</span> <span class="o">=</span> <span class="nf">round</span><span class="p">(</span><span class="n">count</span><span class="p">),</span> <span class="n">colData</span> <span class="o">=</span> <span class="n">coldata</span><span class="p">,</span>
                              <span class="n">design</span> <span class="o">=</span> <span class="o">~</span> <span class="n">condition</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>converting counts to integer mode
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">dds</span> <span class="o">&lt;-</span> <span class="n">dds</span><span class="p">[</span><span class="nf">rowSums</span><span class="p">(</span><span class="nf">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">))</span> <span class="o">&gt;=</span> <span class="m">10</span><span class="p">,]</span>
<span class="n">dds</span><span class="o">$</span><span class="n">condition</span> <span class="o">&lt;-</span> <span class="nf">relevel</span><span class="p">(</span><span class="n">dds</span><span class="o">$</span><span class="n">condition</span><span class="p">,</span> <span class="n">ref</span> <span class="o">=</span> <span class="s">&quot;c&quot;</span><span class="p">)</span>
<span class="n">dds</span> <span class="o">&lt;-</span> <span class="nf">DESeq</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
<span class="nf">resultsNames</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
<span class="n">res</span> <span class="o">&lt;-</span> <span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
<span class="n">res</span> <span class="o">&lt;-</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">res</span><span class="o">$</span><span class="n">padj</span><span class="p">),])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>estimating size factors
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>estimating dispersions
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>gene-wise dispersion estimates
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>mean-dispersion relationship
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>-- note: fitType=&#39;parametric&#39;, but the dispersion trend was not well captured by the
   function: y = a/x + b, and a local regression fit was automatically substituted.
   specify fitType=&#39;local&#39; or &#39;mean&#39; to avoid this message next time.
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>final dispersion estimates
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fitting model and testing
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>-- replacing outliers and refitting for 85 genes
-- DESeq argument &#39;minReplicatesForReplace&#39; = 7 
-- original counts are preserved in counts(dds)
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>estimating dispersions
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fitting model and testing
</pre></div>
</div>
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'Intercept'</li><li>'condition_d_vs_c'</li></ol>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">summary</span><span class="p">(</span><span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">),</span><span class="n">alpha</span> <span class="o">=</span> <span class="m">0.05</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>out of 50892 with nonzero total read count
adjusted p-value &lt; 0.05
LFC &gt; 0 (up)       : 5, 0.0098%
LFC &lt; 0 (down)     : 1, 0.002%
outliers [1]       : 0, 0%
low counts [2]     : 1, 0.002%
(mean count &lt; 0)
[1] see &#39;cooksCutoff&#39; argument of ?results
[2] see &#39;independentFiltering&#39; argument of ?results
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">write.csv</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s">&quot;./data/DESeq2_Results.csv&quot;</span><span class="p">)</span>
<span class="nf">saveRDS</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s">&quot;./data/DESeq2_Results.rds&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="visualization-of-of-differential-expression-results">
<h2>Visualization of of differential expression results.<a class="headerlink" href="#visualization-of-of-differential-expression-results" title="Permalink to this headline">#</a></h2>
<p>To summarise results for all genes visually, mean-difference plots, which display log-FCs from the linear model fit against the average log-CPM values can be generated using the <code class="docutils literal notranslate"><span class="pre">plotMD</span></code> function, with the differentially expressed genes highlighted.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plotMD</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="m">1</span><span class="p">,</span> <span class="n">status</span><span class="o">=</span><span class="n">dt</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span> <span class="n">main</span><span class="o">=</span><span class="nf">colnames</span><span class="p">(</span><span class="n">fit2</span><span class="p">)[</span><span class="m">1</span><span class="p">],</span>
       <span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-8</span><span class="p">,</span><span class="m">13</span><span class="p">),</span><span class="n">pch</span><span class="o">=</span><span class="m">20</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
<span class="nf">abline</span><span class="p">(</span><span class="n">h</span><span class="o">=</span><span class="m">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module02-DE_Analysis_48_0.png" src="_images/Module02-DE_Analysis_48_0.png" />
</div>
</div>
<p>Another way to visualize the number of regulated and unregulated genes is to use Venn diagram. The command to generate the diagram is presented below</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># summarize test results as &quot;up&quot;, &quot;down&quot; or &quot;not expressed&quot;</span>
<span class="n">dT</span> <span class="o">&lt;-</span> <span class="nf">decideTests</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">adjust.method</span><span class="o">=</span><span class="s">&quot;fdr&quot;</span><span class="p">,</span> <span class="n">p.value</span><span class="o">=</span><span class="m">0.05</span><span class="p">)</span>
<span class="c1"># Venn diagram of results</span>
<span class="nf">vennDiagram</span><span class="p">(</span><span class="n">dT</span><span class="p">,</span> <span class="n">circle.col</span><span class="o">=</span><span class="nf">palette</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module02-DE_Analysis_50_0.png" src="_images/Module02-DE_Analysis_50_0.png" />
</div>
</div>
<p>We can visualize the adjusted p-value distribution for all gene using histogram plot as follow</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize and quality control test results.</span>
<span class="c1"># Build histogram of P-values for all genes. Normal test</span>
<span class="c1"># assumption is that most genes are not differentially expressed.</span>
<span class="n">tT2</span> <span class="o">&lt;-</span> <span class="nf">topTable</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">adjust</span><span class="o">=</span><span class="s">&quot;fdr&quot;</span><span class="p">,</span> <span class="n">sort.by</span><span class="o">=</span><span class="s">&quot;B&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="kc">Inf</span><span class="p">)</span>
<span class="nf">hist</span><span class="p">(</span><span class="n">tT2</span><span class="o">$</span><span class="n">adj.P.Val</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="s">&quot;grey&quot;</span><span class="p">,</span> <span class="n">border</span> <span class="o">=</span> <span class="s">&quot;white&quot;</span><span class="p">,</span> <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;P-adj&quot;</span><span class="p">,</span>
     <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;Number of genes&quot;</span><span class="p">,</span> <span class="n">main</span> <span class="o">=</span> <span class="s">&quot;P-adj value distribution&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module02-DE_Analysis_52_0.png" src="_images/Module02-DE_Analysis_52_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># create Q-Q plot for t-statistic</span>
<span class="n">t.good</span> <span class="o">&lt;-</span> <span class="nf">which</span><span class="p">(</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">fit2</span><span class="o">$</span><span class="bp">F</span><span class="p">))</span> <span class="c1"># filter out bad probes</span>
<span class="nf">qqt</span><span class="p">(</span><span class="n">fit2</span><span class="o">$</span><span class="n">t</span><span class="p">[</span><span class="n">t.good</span><span class="p">],</span> <span class="n">fit2</span><span class="o">$</span><span class="n">df.total</span><span class="p">[</span><span class="n">t.good</span><span class="p">],</span> <span class="n">main</span><span class="o">=</span><span class="s">&quot;Moderated t statistic&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module02-DE_Analysis_53_0.png" src="_images/Module02-DE_Analysis_53_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># volcano plot (log P-value vs log fold change)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">fit2</span><span class="p">)</span> <span class="c1"># list contrast names</span>
<span class="n">ct</span> <span class="o">&lt;-</span> <span class="m">1</span>        <span class="c1"># choose contrast of interest</span>
<span class="nf">volcanoplot</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">coef</span><span class="o">=</span><span class="n">ct</span><span class="p">,</span> <span class="n">main</span><span class="o">=</span><span class="nf">colnames</span><span class="p">(</span><span class="n">fit2</span><span class="p">)[</span><span class="n">ct</span><span class="p">],</span> <span class="n">pch</span><span class="o">=</span><span class="m">20</span><span class="p">,</span>
            <span class="n">highlight</span><span class="o">=</span><span class="nf">length</span><span class="p">(</span><span class="nf">which</span><span class="p">(</span><span class="n">dT</span><span class="p">[,</span><span class="n">ct</span><span class="p">]</span><span class="o">!=</span><span class="m">0</span><span class="p">)),</span> <span class="n">names</span><span class="o">=</span><span class="nf">rep</span><span class="p">(</span><span class="s">&#39;+&#39;</span><span class="p">,</span> <span class="nf">nrow</span><span class="p">(</span><span class="n">fit2</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'c-d'</div><img alt="_images/Module02-DE_Analysis_54_1.png" src="_images/Module02-DE_Analysis_54_1.png" />
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            kernelName: "ir",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="Module01-Gene_Mapping.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Gene ID mapping</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="Module03-Gene_Set_and_Pathway.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Gene set and pathway</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Bang Tran, Hung Nguyen, and Tin Nguyen<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>