
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Pathway Analysis &#8212; Consensus Pathway Analysis using Google Cloud Infrastructure</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Ontology and Gene Ontology" href="Module03-Gene_Set_and_Pathway.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/Cover.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Consensus Pathway Analysis using Google Cloud Infrastructure</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Overview
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting Started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="VertexAI.html">
   Setting up Google Cloud environment
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 01
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module01-GEO_Data_Processing.html">
   GEO data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Module01-Gene_Mapping.html">
   Gene ID mapping
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 02
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module02-DE_Analysis.html">
   Differential gene expression analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 03
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module03-Gene_Set_and_Pathway.html">
   Ontology and Gene Ontology
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 04
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Pathway Analysis
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/executablebooks/jupyter-book/master?urlpath=tree/docs/Module04-Pathway_Analysis.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/executablebooks/jupyter-book"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2FModule04-Pathway_Analysis.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/Module04-Pathway_Analysis.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Pathway Analysis
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#learning-objectives">
     Learning Objectives:
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#enrichment-analysis">
     Enrichment analysis
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#enrichment-analysis-using-ora">
     Enrichment analysis using ORA
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#over-representation-analysis-using-go">
     Over Representation Analysis Using GO
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#visualization">
     Visualization
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#enrichment-analysis-using-fgsea">
     Enrichment Analysis using FGSEA
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#gsea-enrichment-analysis">
     GSEA Enrichment Analysis
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#inputs">
   Inputs:
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#steps">
   Steps
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#fgsea-using-go">
   FGSEA Using GO
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-fgsea-analysis-procedure-from-the-author">
     The FGSEA Analysis Procedure (from the author)
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#one-can-make-an-enrichment-plot-for-a-pathway">
   One can make an enrichment plot for a pathway:
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#or-make-a-table-plot-for-a-bunch-of-selected-pathways">
   Or make a table plot for a bunch of selected pathways:
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#fgsea-enrichment-analysis-with-kegg-data">
     FGSEA Enrichment Analysis With KEGG Data
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#fgsea-using-kegg">
   FGSEA Using KEGG
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#to-load-the-gmt-file-for-enrichment-analysis-we-can-use-gsa-read-gmt-function-available-in-the-gsa-r-package-here-is-the-code-to-install-the-package-and-load-the-gmt-file">
   To load the GMT file for enrichment analysis, we can use GSA.read.gmt function available in the GSA R package. Here is the code to install the package and load the GMT file
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-the-gmt-file-from-disk-we-use-invisible-function-to-supress-the-excessive-output-message-from-the-gsa-read-gmt-function">
   Load the GMT file from disk, we use “invisible” function to supress the excessive output message from the “GSA.read.gmt” function
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#view-first-five-pathways-and-related-gene-sets">
   View first five pathways and related gene sets
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#view-the-name-of-the-pathways">
   View the name of the pathways
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#view-the-description-of-each-pathway">
   View the description of each pathway
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#get-p-value-from-de-results">
   Get p-value from DE results
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Pathway Analysis</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Pathway Analysis
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#learning-objectives">
     Learning Objectives:
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#enrichment-analysis">
     Enrichment analysis
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#enrichment-analysis-using-ora">
     Enrichment analysis using ORA
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#over-representation-analysis-using-go">
     Over Representation Analysis Using GO
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#visualization">
     Visualization
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#enrichment-analysis-using-fgsea">
     Enrichment Analysis using FGSEA
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#gsea-enrichment-analysis">
     GSEA Enrichment Analysis
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#inputs">
   Inputs:
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#steps">
   Steps
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#fgsea-using-go">
   FGSEA Using GO
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-fgsea-analysis-procedure-from-the-author">
     The FGSEA Analysis Procedure (from the author)
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#one-can-make-an-enrichment-plot-for-a-pathway">
   One can make an enrichment plot for a pathway:
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#or-make-a-table-plot-for-a-bunch-of-selected-pathways">
   Or make a table plot for a bunch of selected pathways:
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#fgsea-enrichment-analysis-with-kegg-data">
     FGSEA Enrichment Analysis With KEGG Data
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#fgsea-using-kegg">
   FGSEA Using KEGG
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#to-load-the-gmt-file-for-enrichment-analysis-we-can-use-gsa-read-gmt-function-available-in-the-gsa-r-package-here-is-the-code-to-install-the-package-and-load-the-gmt-file">
   To load the GMT file for enrichment analysis, we can use GSA.read.gmt function available in the GSA R package. Here is the code to install the package and load the GMT file
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-the-gmt-file-from-disk-we-use-invisible-function-to-supress-the-excessive-output-message-from-the-gsa-read-gmt-function">
   Load the GMT file from disk, we use “invisible” function to supress the excessive output message from the “GSA.read.gmt” function
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#view-first-five-pathways-and-related-gene-sets">
   View first five pathways and related gene sets
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#view-the-name-of-the-pathways">
   View the name of the pathways
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#view-the-description-of-each-pathway">
   View the description of each pathway
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#get-p-value-from-de-results">
   Get p-value from DE results
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="pathway-analysis">
<h1>Pathway Analysis<a class="headerlink" href="#pathway-analysis" title="Permalink to this headline">#</a></h1>
<p>The main purpose of Pathway Analysis tools is to analyze data obtained from high-throughput technologies, detecting relevant groups of related genes that are altered in case samples in comparison to a control. In this manner, Pathway Analysis methods seek to overcome the problem of interpreting overwhelmingly large lists of important, but isolated genes detached of biological context, which are the main output of most basic high-throughput data analysis, as differential expression analysis. A standard approach to addressing this problem is pathway enrichment analysis, which summarizes the large gene list as a smaller list of more easily interpretable pathways. Pathways are statistically tested for over-representation in the experimental gene list relative to what is expected by chance, using several common statistical tests that consider the number of genes detected in the experiment, their relative ranking and the number of genes annotated to a pathway of interest.</p>
<p>A principled way to interpret such gene sets is to compare them with a database of well-annotated gene sets, such as biological pathways. For example, one of the most widely used approach, Over-representation Analysis (ORA), counts the number of common genes shared by an input gene set and each annotated gene set, and applies a statistical test, such as the cumulative hyper-geometric test, to calculate the statistical significance of the overlap. A p-value cutoff, e.g. 0.05, is then applied to select the annotated gene sets that have statistically significant overlaps with the input gene set. An extension of ORA known as the Gene Set Enrichment Analysis (GSEA) tries to eliminate the need for an ad hoc cutoff (e.g. expression fold change) that is often used in defining the input gene set. GSEA works by ranking all genes in the genome according to, say, level of differential expression, and tests whether any annotated gene set is ranked unexpected high or low through running-sum statistic.</p>
<section id="learning-objectives">
<h2>Learning Objectives:<a class="headerlink" href="#learning-objectives" title="Permalink to this headline">#</a></h2>
<ol class="simple">
<li><p>Introduction to enrichment analysis</p></li>
<li><p>Downloading terms, pathway gene set from GO and KEGG.</p></li>
<li><p>Saving results to GMT file format.</p></li>
</ol>
</section>
<section id="enrichment-analysis">
<h2>Enrichment analysis<a class="headerlink" href="#enrichment-analysis" title="Permalink to this headline">#</a></h2>
<p>Gene Set Enrichment Analysis methods help gain insight into obtained gene lists by identifying pathways that are enriched in a gene list more than would be expected by chance. The three major steps involved the process include; definition of a gene list from omics data, determination of statistically enriched pathways, and visualization and interpretation of the results.</p>
</section>
<section id="enrichment-analysis-using-ora">
<h2>Enrichment analysis using ORA<a class="headerlink" href="#enrichment-analysis-using-ora" title="Permalink to this headline">#</a></h2>
<p>Over-representation (or enrichment) analysis is a statistical method that determines whether genes from pre-defined sets (ex: those belonging to a specific GO term or KEGG pathway) are present more than would be expected (over-represented) in a subset of your data.</p>
<p>The p-value can be calculated by hypergeometric distribution.</p>
<p><img alt="" src="_images/ora_p_value.png" /></p>
<p>where N is the total number of genes in the background distribution, M is the number of genes within that distribution that are annotated (either directly or indirectly) to the gene set of interest, n is the size of the list of genes of interest and k is the number of genes within that list which are annotated to the gene set. The background distribution by default is all the genes that have annotation.</p>
</section>
<section id="over-representation-analysis-using-go">
<h2>Over Representation Analysis Using GO<a class="headerlink" href="#over-representation-analysis-using-go" title="Permalink to this headline">#</a></h2>
<p>library(tidyverse)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">&lt;-</span> <span class="nf">read_csv</span><span class="p">(</span><span class="s">&quot;./data/DE_genes.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="n">Error</span> <span class="ow">in</span> <span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;./data/DE_genes.csv&quot;</span><span class="p">):</span> <span class="n">could</span> <span class="ow">not</span> <span class="n">find</span> <span class="n">function</span> <span class="s2">&quot;read_csv&quot;</span>
<span class="ne">Traceback</span>:
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">data.head</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">mask</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">$</span><span class="n">adj.P.Val</span> <span class="o">&lt;</span> <span class="m">0.05</span> <span class="o">&amp;</span>
        <span class="nf">abs</span><span class="p">(</span><span class="n">data</span><span class="o">$</span><span class="n">logFC</span><span class="p">)</span> <span class="o">&gt;</span> <span class="nf">log2</span><span class="p">(</span><span class="m">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">deGenes</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">mask</span><span class="p">,])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">deGenes</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">length</span><span class="p">(</span><span class="n">deGenes</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">geneUniverse</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">length</span><span class="p">(</span><span class="n">geneUniverse</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>In order to assess functional enrichment, both DE gene list and gene universe must be annotated in Entrez IDs:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">deGenes</span> <span class="o">&lt;-</span> <span class="nf">unlist</span><span class="p">(</span><span class="nf">mget</span><span class="p">(</span><span class="n">deGenes</span><span class="p">,</span> <span class="n">envir</span><span class="o">=</span><span class="n">org.Hs.egENSEMBL2EG</span><span class="p">,</span>
                       <span class="n">ifnotfound</span> <span class="o">=</span> <span class="kc">NA</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">geneUniverse</span> <span class="o">&lt;-</span> <span class="nf">unlist</span><span class="p">(</span><span class="nf">mget</span><span class="p">(</span><span class="n">geneUniverse</span><span class="p">,</span> <span class="n">envir</span><span class="o">=</span><span class="n">org.Hs.egENSEMBL2EG</span><span class="p">,</span>
                            <span class="n">ifnotfound</span> <span class="o">=</span> <span class="kc">NA</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<p>The GO enrichment analysis using clusterProfiler is performed by</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">clusterProfiler</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ans.go</span> <span class="o">&lt;-</span> <span class="nf">enrichGO</span><span class="p">(</span><span class="n">gene</span> <span class="o">=</span> <span class="n">deGenes</span><span class="p">,</span> <span class="n">ont</span> <span class="o">=</span> <span class="s">&quot;BP&quot;</span><span class="p">,</span>
                   <span class="n">OrgDb</span> <span class="o">=</span><span class="s">&quot;org.Hs.eg.db&quot;</span><span class="p">,</span>
                   <span class="n">universe</span> <span class="o">=</span> <span class="n">geneUniverse</span><span class="p">,</span>
                   <span class="n">readable</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span>
                   <span class="n">pvalueCutoff</span> <span class="o">=</span> <span class="m">0.05</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">tab.go</span> <span class="o">&lt;-</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="n">ans.go</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">tab.go</span><span class="o">&lt;-</span> <span class="nf">subset</span><span class="p">(</span><span class="n">tab.go</span><span class="p">,</span> <span class="n">Count</span><span class="o">&gt;</span><span class="m">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">tab.go</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="visualization">
<h2>Visualization<a class="headerlink" href="#visualization" title="Permalink to this headline">#</a></h2>
<p>All analyses performed with clusterProfiler can be visualized by different plots:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">enrichplot</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p1</span> <span class="o">&lt;-</span> <span class="nf">barplot</span><span class="p">(</span><span class="n">ans.dis</span><span class="p">,</span> <span class="n">showCategory</span><span class="o">=</span><span class="m">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p1</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p2</span> <span class="o">&lt;-</span> <span class="nf">dotplot</span><span class="p">(</span><span class="n">ans.kegg</span><span class="p">,</span> <span class="n">showCategory</span><span class="o">=</span><span class="m">20</span><span class="p">)</span> <span class="o">+</span> <span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;KEGG&quot;</span><span class="p">)</span>
<span class="n">p3</span> <span class="o">&lt;-</span> <span class="nf">dotplot</span><span class="p">(</span><span class="n">ans.dis</span><span class="p">,</span> <span class="n">showCategory</span><span class="o">=</span><span class="m">20</span><span class="p">)</span> <span class="o">+</span> <span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;Disease&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot_grid</span><span class="p">(</span><span class="n">p2</span><span class="p">,</span> <span class="n">p3</span><span class="p">,</span> <span class="n">nrow</span><span class="o">=</span><span class="m">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p4</span> <span class="o">&lt;-</span> <span class="nf">upsetplot</span><span class="p">(</span><span class="n">ans.dis</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p4</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p5</span> <span class="o">&lt;-</span> <span class="nf">emapplot</span><span class="p">(</span><span class="n">ans.kegg</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p5</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span> <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;cowplot&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cowplot</span><span class="o">::</span><span class="nf">plot_grid</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span> <span class="n">p3</span><span class="p">,</span> <span class="n">p5</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="m">2</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="kc">LETTERS</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="enrichment-analysis-using-fgsea">
<h2>Enrichment Analysis using FGSEA<a class="headerlink" href="#enrichment-analysis-using-fgsea" title="Permalink to this headline">#</a></h2>
<p>This submodule describes FGSEA, one of the tools for evaluating pathway enrichment in transcriptional data and it stands for Fast preranked Gene Set Enrichment Analysis (GSEA). It can quickly and accurately calculate arbitrarily low GSEA P-values for a collection of gene sets based on an adaptive multi-level split Monte-Carlo scheme.</p>
<p>Typical experimental design consists in comparing two conditions with several replicates using a differential gene expression test followed by preranked GSEA performed against a collection of hundreds and thousands of pathways. However, the reference implementation of this method cannot accurately estimate small P-values, which significantly limits its sensitivity due to multiple hypotheses correction procedure. FGSEA, on the other hand, is able to estimate arbitrarily low GSEA P-values with a high accuracy in a matter of minutes or even seconds.</p>
</section>
<section id="gsea-enrichment-analysis">
<h2>GSEA Enrichment Analysis<a class="headerlink" href="#gsea-enrichment-analysis" title="Permalink to this headline">#</a></h2>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="inputs">
<h1>Inputs:<a class="headerlink" href="#inputs" title="Permalink to this headline">#</a></h1>
<p>gene_list = Ranked gene list ( numeric vector, names of vector should be gene names)
GO_file= Path to the “gmt” GO file on your system.
pval = P-value threshold for returning results</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="steps">
<h1>Steps<a class="headerlink" href="#steps" title="Permalink to this headline">#</a></h1>
<ol class="simple">
<li><p>Run GSEA (package: fgsea)</p></li>
<li><p>Collapse redundant GO terms using a permutation test</p></li>
<li><p>Return GSEA plot and data.frame of results</p></li>
</ol>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="fgsea-using-go">
<h1>FGSEA Using GO<a class="headerlink" href="#fgsea-using-go" title="Permalink to this headline">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">&lt;-</span> <span class="nf">read_csv</span><span class="p">(</span><span class="s">&quot;./data/DE_genes.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">data</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">mask</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">$</span><span class="n">adj.P.Val</span> <span class="o">&lt;</span> <span class="m">0.05</span> <span class="o">&amp;</span>
  <span class="nf">abs</span><span class="p">(</span><span class="n">data</span><span class="o">$</span><span class="n">logFC</span><span class="p">)</span> <span class="o">&gt;</span> <span class="nf">log2</span><span class="p">(</span><span class="m">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">deGenes</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">mask</span><span class="p">,])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">deGenes</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">length</span><span class="p">(</span><span class="n">deGenes</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">geneUniverse</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">length</span><span class="p">(</span><span class="n">geneUniverse</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="the-fgsea-analysis-procedure-from-the-author">
<h2>The FGSEA Analysis Procedure (from the author)<a class="headerlink" href="#the-fgsea-analysis-procedure-from-the-author" title="Permalink to this headline">#</a></h2>
<p>Loading Libraries</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">fgsea</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Loading example pathways and gene-level statistics:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">data</span><span class="p">(</span><span class="n">examplePathways</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">data</span><span class="p">(</span><span class="n">exampleRanks</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>#fgsea has a default lower bound eps=1e-10 for estimating P-values. If you need to estimate P-value more accurately, you can set the eps argument to zero in the fgsea function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">fgseaRes</span> <span class="o">&lt;-</span> <span class="nf">fgsea</span><span class="p">(</span><span class="n">pathways</span> <span class="o">=</span> <span class="n">examplePathways</span><span class="p">,</span>
                  <span class="n">stats</span>    <span class="o">=</span> <span class="n">exampleRanks</span><span class="p">,</span>
                  <span class="n">eps</span>      <span class="o">=</span> <span class="m">0.0</span><span class="p">,</span>
                  <span class="n">minSize</span>  <span class="o">=</span> <span class="m">15</span><span class="p">,</span>
                  <span class="n">maxSize</span>  <span class="o">=</span> <span class="m">500</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">head</span><span class="p">(</span><span class="n">fgseaRes</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">pval</span><span class="p">),</span> <span class="p">])</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="one-can-make-an-enrichment-plot-for-a-pathway">
<h1>One can make an enrichment plot for a pathway:<a class="headerlink" href="#one-can-make-an-enrichment-plot-for-a-pathway" title="Permalink to this headline">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plotEnrichment</span><span class="p">(</span><span class="n">examplePathways</span><span class="p">[[</span><span class="s">&quot;5991130_Programmed_Cell_Death&quot;</span><span class="p">]],</span>
               <span class="n">exampleRanks</span><span class="p">)</span> <span class="o">+</span> <span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s">&quot;Programmed Cell Death&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="or-make-a-table-plot-for-a-bunch-of-selected-pathways">
<h1>Or make a table plot for a bunch of selected pathways:<a class="headerlink" href="#or-make-a-table-plot-for-a-bunch-of-selected-pathways" title="Permalink to this headline">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">topPathwaysUp</span> <span class="o">&lt;-</span> <span class="n">fgseaRes</span><span class="p">[</span><span class="n">ES</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">][</span><span class="nf">head</span><span class="p">(</span><span class="nf">order</span><span class="p">(</span><span class="n">pval</span><span class="p">),</span> <span class="n">n</span><span class="o">=</span><span class="m">10</span><span class="p">),</span> <span class="n">pathway</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">topPathwaysDown</span> <span class="o">&lt;-</span> <span class="n">fgseaRes</span><span class="p">[</span><span class="n">ES</span> <span class="o">&lt;</span> <span class="m">0</span><span class="p">][</span><span class="nf">head</span><span class="p">(</span><span class="nf">order</span><span class="p">(</span><span class="n">pval</span><span class="p">),</span> <span class="n">n</span><span class="o">=</span><span class="m">10</span><span class="p">),</span> <span class="n">pathway</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">topPathways</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="n">topPathwaysUp</span><span class="p">,</span> <span class="nf">rev</span><span class="p">(</span><span class="n">topPathwaysDown</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plotGseaTable</span><span class="p">(</span><span class="n">examplePathways</span><span class="p">[</span><span class="n">topPathways</span><span class="p">],</span> <span class="n">exampleRanks</span><span class="p">,</span> <span class="n">fgseaRes</span><span class="p">,</span>
              <span class="n">gseaParam</span><span class="o">=</span><span class="m">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="fgsea-enrichment-analysis-with-kegg-data">
<h2>FGSEA Enrichment Analysis With KEGG Data<a class="headerlink" href="#fgsea-enrichment-analysis-with-kegg-data" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># To load the GMT file for enrichment analysis, we can use GSA.read.gmt function available in the GSA R package. Here is the code to install the package and load the GMT file</span>

<span class="nf">suppressMessages</span><span class="p">({</span>
  <span class="nf">suppressWarnings</span><span class="p">(</span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;GSA&quot;</span><span class="p">))</span>
  <span class="nf">suppressPackageStartupMessages</span><span class="p">({</span>
    <span class="nf">suppressWarnings</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="n">GSA</span><span class="p">))</span>
  <span class="p">})</span>
<span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the GMT file from disk, we use &quot;invisible&quot; function to supress the excessive output message from the &quot;GSA.read.gmt&quot; function</span>

<span class="nf">invisible</span><span class="p">(</span><span class="nf">capture.output</span><span class="p">(</span><span class="n">pathways</span> <span class="o">&lt;-</span> <span class="n">GSA</span><span class="o">::</span><span class="nf">GSA.read.gmt</span><span class="p">(</span><span class="s">&quot;./data/KEGG_pathways.gmt&quot;</span><span class="p">)))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># View first five pathways and related gene sets</span>
<span class="n">pathways</span><span class="o">$</span><span class="n">genesets</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># View the name of the pathways</span>
<span class="n">pathways</span><span class="o">$</span><span class="n">geneset.names</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># View the description of each pathway</span>
<span class="n">pathways</span><span class="o">$</span><span class="n">geneset.descriptions</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="fgsea-using-kegg">
<h1>FGSEA Using KEGG<a class="headerlink" href="#fgsea-using-kegg" title="Permalink to this headline">#</a></h1>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="to-load-the-gmt-file-for-enrichment-analysis-we-can-use-gsa-read-gmt-function-available-in-the-gsa-r-package-here-is-the-code-to-install-the-package-and-load-the-gmt-file">
<h1>To load the GMT file for enrichment analysis, we can use GSA.read.gmt function available in the GSA R package. Here is the code to install the package and load the GMT file<a class="headerlink" href="#to-load-the-gmt-file-for-enrichment-analysis-we-can-use-gsa-read-gmt-function-available-in-the-gsa-r-package-here-is-the-code-to-install-the-package-and-load-the-gmt-file" title="Permalink to this headline">#</a></h1>
<p>suppressMessages({
suppressWarnings(install.packages(“GSA”))
suppressPackageStartupMessages({
suppressWarnings(library(GSA))
})
})</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="load-the-gmt-file-from-disk-we-use-invisible-function-to-supress-the-excessive-output-message-from-the-gsa-read-gmt-function">
<h1>Load the GMT file from disk, we use “invisible” function to supress the excessive output message from the “GSA.read.gmt” function<a class="headerlink" href="#load-the-gmt-file-from-disk-we-use-invisible-function-to-supress-the-excessive-output-message-from-the-gsa-read-gmt-function" title="Permalink to this headline">#</a></h1>
<p>invisible(capture.output(pathways &lt;- GSA::GSA.read.gmt(“./data/KEGG_pathways.gmt”)))
pathways</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="view-first-five-pathways-and-related-gene-sets">
<h1>View first five pathways and related gene sets<a class="headerlink" href="#view-first-five-pathways-and-related-gene-sets" title="Permalink to this headline">#</a></h1>
<p>pathways$genesets[1:5]</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="view-the-name-of-the-pathways">
<h1>View the name of the pathways<a class="headerlink" href="#view-the-name-of-the-pathways" title="Permalink to this headline">#</a></h1>
<p>pathways$geneset.names[1:5]</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="view-the-description-of-each-pathway">
<h1>View the description of each pathway<a class="headerlink" href="#view-the-description-of-each-pathway" title="Permalink to this headline">#</a></h1>
<p>pathways$geneset.descriptions[1:5]</p>
<p>#//GeneList
DE &lt;- readRDS(“./data/DE_genes.rds”)</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="get-p-value-from-de-results">
<h1>Get p-value from DE results<a class="headerlink" href="#get-p-value-from-de-results" title="Permalink to this headline">#</a></h1>
<p>gene_list &lt;- DE$adj.P.Val
#// Gene List
gene_list
GSEA = function(gene_list, pathways, pval) {
set.seed(54321)</p>
<p>library(dplyr)
library(fgsea)
if ( any( duplicated(names(gene_list)) )  ) {
warning(“Duplicates in gene names”)
gene_list = gene_list[!duplicated(names(gene_list))]
}
if  ( !all( order(gene_list, decreasing = TRUE) == 1:length(gene_list)) ){
warning(“Gene list not sorted”)
gene_list = sort(gene_list, decreasing = TRUE)
}</p>
<p>myGO = fgsea::gmtPathways(pathways)</p>
<p>fgRes &lt;- fgsea::fgsea(pathways = myGO,
stats = gene_list,
minSize=15, ## minimum gene set size
maxSize=400, ## maximum gene set size
nperm=10000) %&gt;%
as.data.frame() %&gt;%
dplyr::filter(padj &lt; !!pval) %&gt;%
arrange(desc(NES))
message(paste(“Number of signficant gene sets =”, nrow(fgRes)))</p>
<p>message(“Collapsing Pathways —–“)
concise_pathways = collapsePathways(data.table::as.data.table(fgRes),
pathways = myGO,
stats = gene_list)
fgRes = fgRes[fgRes<span class="math notranslate nohighlight">\(pathway %in% concise_pathways\)</span>mainPathways, ]
message(paste(“Number of gene sets after collapsing =”, nrow(fgRes)))</p>
<p>fgRes<span class="math notranslate nohighlight">\(Enrichment = ifelse(fgRes\)</span>NES &gt; 0, “Up-regulated”, “Down-regulated”)
filtRes = rbind(head(fgRes, n = 10),
tail(fgRes, n = 10 ))</p>
<p>total_up = sum(fgRes<span class="math notranslate nohighlight">\(Enrichment == &quot;Up-regulated&quot;)
  total_down = sum(fgRes\)</span>Enrichment == “Down-regulated”)
header = paste0(“Top 10 (Total pathways: Up=”, total_up,”, Down=”,    total_down, “)”)</p>
<p>colos = setNames(c(“firebrick2”, “dodgerblue2”),
c(“Up-regulated”, “Down-regulated”))</p>
<p>g1= ggplot(filtRes, aes(reorder(pathway, NES), NES)) +
geom_point( aes(fill = Enrichment, size = size), shape=21) +
scale_fill_manual(values = colos ) +
scale_size_continuous(range = c(2,10)) +
geom_hline(yintercept = 0) +
coord_flip() +
labs(x=”Pathway”, y=”Normalized Enrichment Score”,
title=header) +
th</p>
<p>output = list(“Results” = fgRes, “Plot” = g1)
return(output)
}</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">## Enrichment analysis using gsa</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># // GSA using GO</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># // GSA with KEGG</span>
</pre></div>
</div>
</div>
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            kernelName: "ir",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="Module03-Gene_Set_and_Pathway.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Ontology and Gene Ontology</p>
        </div>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Bang Tran, Hung Nguyen, and Tin Nguyen<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>