
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Gene ID mapping &#8212; Consensus Pathway Analysis using Google Cloud Infrastructure</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Differential gene expression analysis" href="Module02-DE_Analysis.html" />
    <link rel="prev" title="GEO data processing" href="Module01-GEO_Data_Processing.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/Cover.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Consensus Pathway Analysis using Google Cloud Infrastructure</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Overview
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting Started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="VertexAI.html">
   Setting up Google Cloud environment
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 01
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module01-GEO_Data_Processing.html">
   GEO data processing
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Gene ID mapping
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 02
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module02-DE_Analysis.html">
   Differential gene expression analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 03
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module03-Gene_Set_and_Pathway.html">
   Gene set and pathway
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 04
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module04-Pathway_Analysis.html">
   Pathway Analysis
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/executablebooks/jupyter-book/master?urlpath=tree/docs/Module01-Gene_Mapping.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/executablebooks/jupyter-book"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2FModule01-Gene_Mapping.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/Module01-Gene_Mapping.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#learning-objectives">
   Learning Objectives:
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#understanding-different-probe-set-id">
     Understanding different probe set ID
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#mapping-probe-ids-into-gene-identifiers-and-symbols">
     Mapping probe IDs into gene identifiers and symbols.
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Gene ID mapping</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#learning-objectives">
   Learning Objectives:
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#understanding-different-probe-set-id">
     Understanding different probe set ID
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#mapping-probe-ids-into-gene-identifiers-and-symbols">
     Mapping probe IDs into gene identifiers and symbols.
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="gene-id-mapping">
<h1>Gene ID mapping<a class="headerlink" href="#gene-id-mapping" title="Permalink to this headline">#</a></h1>
<p><img alt="" src="_images/Gene_ID_Conversion.jpg" /></p>
<section id="learning-objectives">
<h2>Learning Objectives:<a class="headerlink" href="#learning-objectives" title="Permalink to this headline">#</a></h2>
<ol class="simple">
<li><p>Understanding different probe set ID.</p></li>
<li><p>Mapping probe IDs into gene identifiers and symbols.</p></li>
</ol>
<section id="understanding-different-probe-set-id">
<h3>Understanding different probe set ID<a class="headerlink" href="#understanding-different-probe-set-id" title="Permalink to this headline">#</a></h3>
<p>Gene set or pathway analysis requires that gene sets and expression data use the same type of gene ID
(Entrez Gene ID, Gene symbol or probe set ID etc). However, this is frequently not true for the data we have.
For example, our gene sets mostly use Entrez Gene ID, but microarray datasets are frequently labeled by probe
set ID (or RefSeq transcript ID etc). Therefore, we need to convert or map the probe set IDs to Entrez gene ID.
Here, we will use <code class="docutils literal notranslate"><span class="pre">GSE48350</span></code> dataset that we have used in the previous section for demonstration of gene ID mapping.
In order to know what kind of probe set ID, we need to navigate to GEO record page of <code class="docutils literal notranslate"><span class="pre">GSE48350</span></code>. Under the <code class="docutils literal notranslate"><span class="pre">Platform</span></code>
tab, we can find the probe ID information.
<img alt="probID" src="_images/ProbeID.png" /></p>
<p>From the record page, we know that the dataset was generated from 1 platform using Affymetrix Human Genome U133 Plus 2.0 Array.
To convert or map the probe set IDs to Entrez gene ID, we need to find the corresponding annotation package from
<a href="https://bioconductor.org/">Bioconductor</a>. For analyzed data, we need to use the <code class="docutils literal notranslate"><span class="pre">hgu133plus2.db</span></code> database.
We can install the package using following R command:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">suppressMessages</span><span class="p">({</span><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
    <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
<span class="nf">suppressWarnings</span><span class="p">(</span><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;hgu133plus2.db&quot;</span><span class="p">,</span> <span class="n">update</span> <span class="o">=</span> <span class="bp">F</span><span class="p">))</span>
<span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<p>In the previous section, we have successfully downloaded the dataset <code class="docutils literal notranslate"><span class="pre">GSE48350</span></code> and saved a sub-set of it to the <code class="docutils literal notranslate"><span class="pre">data</span></code> sub-directory.
Now we can load the sample dataset and check for the Probe Set ID names by using the following command.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="s">&quot;./data/GSE48350.rds&quot;</span><span class="p">)</span>
<span class="n">probeIDs</span> <span class="o">=</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">probeIDs</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'1007_s_at'</li><li>'1053_at'</li><li>'117_at'</li><li>'121_at'</li><li>'1255_g_at'</li><li>'1294_at'</li><li>'1316_at'</li><li>'1320_at'</li><li>'1405_i_at'</li><li>'1431_at'</li></ol>
</div></div>
</div>
<p><strong>Probe Set ID</strong> is the identifier that refers to a set of probe pairs selected to represent expressed sequences on an array. Designations are given at design time.</p>
<p>The probe set names never change, but they can give you an idea of what was known about the sequence at the time of design.</p>
<ul class="simple">
<li><p><strong>_at</strong> = all the probes hit one known transcript.</p></li>
<li><p><strong>_a</strong> = all probes in the set hit alternate transcripts from the same genes.</p></li>
<li><p><strong>_s</strong> = all probes in the set hit transcripts from different genes.</p></li>
<li><p><strong>_x</strong> = some probes hit transcripts from different genes.</p></li>
</ul>
<p>For HG-U133, the <strong>_a</strong> designation was not used; an <strong>_s</strong> probe set on these arrays means the same as an <strong>_a</strong> on any of the HG-U133 arrays.</p>
</section>
<section id="mapping-probe-ids-into-gene-identifiers-and-symbols">
<h3>Mapping probe IDs into gene identifiers and symbols.<a class="headerlink" href="#mapping-probe-ids-into-gene-identifiers-and-symbols" title="Permalink to this headline">#</a></h3>
<p>For pathway analysis, Probe set IDs should be mapped to Ensembl ID and SYMBOL.
Ensembl ID is a transcript identifier from the ENSEMBL project while SYMBOL is a gene symbol is a short-form abbreviation for a particular gene.
We can load the <code class="docutils literal notranslate"><span class="pre">hgu133plus2.db</span></code> database to perform Probe ID mapping using the following command:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">suppressMessages</span><span class="p">({</span><span class="nf">library</span><span class="p">(</span><span class="n">hgu133plus2.db</span><span class="p">)})</span>
</pre></div>
</div>
</div>
</div>
<p>Here, we can build a  lookup table that maps the Probe ID to Ensembl ID and gene symbol using the following command:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">suppressMessages</span><span class="p">({</span>
<span class="n">annotLookup</span> <span class="o">&lt;-</span> <span class="nf">select</span><span class="p">(</span><span class="n">hgu133plus2.db</span><span class="p">,</span> <span class="n">keys</span> <span class="o">=</span> <span class="n">probeIDs</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;PROBEID&#39;</span><span class="p">,</span> <span class="s">&#39;ENSEMBL&#39;</span><span class="p">,</span> <span class="s">&#39;SYMBOL&#39;</span><span class="p">))</span>
<span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">annotLookup</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 17 × 3</caption>
<thead>
	<tr><th scope=col>PROBEID</th><th scope=col>ENSEMBL</th><th scope=col>SYMBOL</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><td>1007_s_at</td><td>ENSG00000204580</td><td>DDR1  </td></tr>
	<tr><td>1007_s_at</td><td>ENSG00000223680</td><td>DDR1  </td></tr>
	<tr><td>1007_s_at</td><td>ENSG00000229767</td><td>DDR1  </td></tr>
	<tr><td>1007_s_at</td><td>ENSG00000234078</td><td>DDR1  </td></tr>
	<tr><td>1007_s_at</td><td>ENSG00000215522</td><td>DDR1  </td></tr>
	<tr><td>1007_s_at</td><td>ENSG00000137332</td><td>DDR1  </td></tr>
	<tr><td>1007_s_at</td><td>ENSG00000230456</td><td>DDR1  </td></tr>
	<tr><td>1053_at  </td><td>ENSG00000049541</td><td>RFC2  </td></tr>
	<tr><td>117_at   </td><td>ENSG00000173110</td><td>HSPA6 </td></tr>
	<tr><td>121_at   </td><td>ENSG00000125618</td><td>PAX8  </td></tr>
	<tr><td>1255_g_at</td><td>ENSG00000048545</td><td>GUCA1A</td></tr>
	<tr><td>1294_at  </td><td>ENSG00000182179</td><td>UBA7  </td></tr>
	<tr><td>1316_at  </td><td>ENSG00000126351</td><td>THRA  </td></tr>
	<tr><td>1320_at  </td><td>ENSG00000070778</td><td>PTPN21</td></tr>
	<tr><td>1405_i_at</td><td>ENSG00000271503</td><td>CCL5  </td></tr>
	<tr><td>1405_i_at</td><td>ENSG00000274233</td><td>CCL5  </td></tr>
	<tr><td>1431_at  </td><td>ENSG00000130649</td><td>CYP2E1</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>From the lookup table we can spot that a single Probe ID can be mapped to multiple Ensembl IDs.
Depending on analysis performed, users can select their own Ensembl ID of interest. Here is a simple way to
replace Probe IDs with gene symbols :</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">rownames</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">unique</span><span class="p">(</span><span class="n">annotLookup</span><span class="o">$</span><span class="n">SYMBOL</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 10 × 10</caption>
<thead>
	<tr><th></th><th scope=col>GSM300166</th><th scope=col>GSM300167</th><th scope=col>GSM300168</th><th scope=col>GSM300169</th><th scope=col>GSM300170</th><th scope=col>GSM300171</th><th scope=col>GSM300172</th><th scope=col>GSM300173</th><th scope=col>GSM300174</th><th scope=col>GSM300175</th></tr>
	<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>DDR1</th><td>0.9168711</td><td>1.2842289</td><td>1.3838295</td><td>1.4620315</td><td>1.4960686</td><td>1.2866872</td><td>1.3716376</td><td>0.6085351</td><td>0.8452452</td><td>0.6745393</td></tr>
	<tr><th scope=row>RFC2</th><td>0.7367870</td><td>0.9152165</td><td>1.5155498</td><td>1.0262097</td><td>1.2758373</td><td>1.3132664</td><td>1.1284863</td><td>0.9081397</td><td>0.9393891</td><td>1.0130529</td></tr>
	<tr><th scope=row>HSPA6</th><td>0.8950221</td><td>1.0440652</td><td>2.0347068</td><td>1.6976205</td><td>2.5993895</td><td>3.8876049</td><td>1.6831516</td><td>0.8674964</td><td>1.4063434</td><td>0.7850503</td></tr>
	<tr><th scope=row>PAX8</th><td>0.9819619</td><td>1.0361480</td><td>0.9871635</td><td>1.1363467</td><td>1.0282362</td><td>0.8330552</td><td>1.0085661</td><td>1.1053528</td><td>0.9709147</td><td>0.9327358</td></tr>
	<tr><th scope=row>GUCA1A</th><td>0.5765267</td><td>0.6206102</td><td>2.0633803</td><td>1.3301512</td><td>1.6716562</td><td>2.1832146</td><td>1.6839774</td><td>1.2764652</td><td>2.5877041</td><td>1.2204636</td></tr>
	<tr><th scope=row>UBA7</th><td>0.8753293</td><td>0.9479945</td><td>1.3330627</td><td>1.3545021</td><td>1.4012277</td><td>1.0190251</td><td>1.7229316</td><td>0.8337133</td><td>1.0038313</td><td>0.9117344</td></tr>
	<tr><th scope=row>THRA</th><td>1.4066110</td><td>1.0097321</td><td>0.6604550</td><td>0.6860092</td><td>0.7863153</td><td>0.5898622</td><td>0.9050921</td><td>1.1971904</td><td>0.8503563</td><td>0.9668145</td></tr>
	<tr><th scope=row>PTPN21</th><td>0.9490653</td><td>0.8197735</td><td>0.5816383</td><td>0.5428258</td><td>0.4708366</td><td>0.5567084</td><td>0.6201315</td><td>0.8163207</td><td>1.0518593</td><td>0.9359782</td></tr>
	<tr><th scope=row>CCL5</th><td>1.3552448</td><td>0.8693293</td><td>1.4550548</td><td>1.1130387</td><td>0.9304383</td><td>1.0561844</td><td>1.6140728</td><td>1.0406454</td><td>1.1045746</td><td>0.8406106</td></tr>
	<tr><th scope=row>CYP2E1</th><td>1.1783971</td><td>0.8985217</td><td>1.9054227</td><td>1.4497262</td><td>1.4704667</td><td>1.5978145</td><td>1.1633086</td><td>0.9495343</td><td>0.9399777</td><td>1.3396635</td></tr>
</tbody>
</table>
</div></div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            kernelName: "ir",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="Module01-GEO_Data_Processing.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">GEO data processing</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="Module02-DE_Analysis.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Differential gene expression analysis</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Bang Tran, Hung Nguyen, and Tin Nguyen<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>