
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Consensus Pathway Analysis using Google Cloud Infrastructure &#8212; Consensus Pathway Analysis using Google Cloud Infrastructure</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/Cover.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Consensus Pathway Analysis using Google Cloud Infrastructure</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Overview
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting Started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="VertexAI.html">
   Setting up Google Cloud environment
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 01
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module01-GEO_Data_Processing.html">
   GEO data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Module01-Gene_Mapping.html">
   Gene ID mapping
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 02
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module02-DE_Analysis.html">
   Differential gene expression analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 03
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module03-Gene_Set_and_Pathway.html">
   Gene set and pathway
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 04
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module04-Pathway_Analysis.html">
   Pathway Analysis
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/executablebooks/jupyter-book"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2FREADME.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/README.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.md</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Consensus Pathway Analysis using Google Cloud Infrastructure
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#overview-of-page-contents">
     Overview of Page Contents
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#a-name-gs-getting-started-a">
     <a name="GS">
      Getting Started
     </a>
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#a-name-ov-overview-a">
     <a name="OV">
      Overview
     </a>
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#software-requirements-a-name-sof-a">
     <strong>
      Software Requirements
     </strong>
     <a name="SOF">
     </a>
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#workflow-diagrams-a-name-work-a">
     <strong>
      Workflow Diagrams
     </strong>
     <a name="WORK">
     </a>
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#img-src-images-intro-main-img-png-width-900-height-700">
   <img height="700" src="./images/Intro/Main-img.png" width="900"/>
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data-a-name-data-a">
     <strong>
      Data
     </strong>
     <a name="DATA">
     </a>
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#troubleshooting-a-name-tr-a">
     <strong>
      Troubleshooting
     </strong>
     <a name="TR">
     </a>
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#funding-a-name-fund-a">
     <strong>
      Funding
     </strong>
     <a name="FUND">
     </a>
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#optional-but-recommended-license-for-data-a-name-lic-a">
     <strong>
      (Optional, but recommended) License for Data
     </strong>
     <a name="LIC">
     </a>
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Consensus Pathway Analysis using Google Cloud Infrastructure</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Consensus Pathway Analysis using Google Cloud Infrastructure
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#overview-of-page-contents">
     Overview of Page Contents
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#a-name-gs-getting-started-a">
     <a name="GS">
      Getting Started
     </a>
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#a-name-ov-overview-a">
     <a name="OV">
      Overview
     </a>
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#software-requirements-a-name-sof-a">
     <strong>
      Software Requirements
     </strong>
     <a name="SOF">
     </a>
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#workflow-diagrams-a-name-work-a">
     <strong>
      Workflow Diagrams
     </strong>
     <a name="WORK">
     </a>
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#img-src-images-intro-main-img-png-width-900-height-700">
   <img height="700" src="./images/Intro/Main-img.png" width="900"/>
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data-a-name-data-a">
     <strong>
      Data
     </strong>
     <a name="DATA">
     </a>
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#troubleshooting-a-name-tr-a">
     <strong>
      Troubleshooting
     </strong>
     <a name="TR">
     </a>
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#funding-a-name-fund-a">
     <strong>
      Funding
     </strong>
     <a name="FUND">
     </a>
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#optional-but-recommended-license-for-data-a-name-lic-a">
     <strong>
      (Optional, but recommended) License for Data
     </strong>
     <a name="LIC">
     </a>
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="consensus-pathway-analysis-using-google-cloud-infrastructure">
<h1>Consensus Pathway Analysis using Google Cloud Infrastructure<a class="headerlink" href="#consensus-pathway-analysis-using-google-cloud-infrastructure" title="Permalink to this headline">#</a></h1>
<p>This cloud-based learning module teaches pathway analyses using omics datasets to discover consistent
biological mechanism behind a condition.Pathway analysis refers to a set of widely used tools for research in life sciences
intended to give meaning to high-throughput biological data.The main purpose of PA tools is to analyze these data,
detecting relevant groups of related genes that are altered in case samples in comparison to a control.
Pathway analysis approaches use available pathway databases and the given gene expression data to identify the pathways
which are significantly impacted in a given condition.</p>
<p>The course is structured such that the content will be arranged in five sub-modules which allows us to:</p>
<ol class="simple">
<li><p>Download and process data from public repositories,</p></li>
<li><p>Perform differential analysis,</p></li>
<li><p>Perform pathway analysis using different methods that seek to answer different research hypotheses,</p></li>
<li><p>Perform meta-analysis and combine methods and datasets to find consensus results, and</p></li>
<li><p>Interactively explore significantly impacted pathways across multiple analyses, and browsing relationships between
pathways and genes.</p></li>
</ol>
<section id="overview-of-page-contents">
<h2>Overview of Page Contents<a class="headerlink" href="#overview-of-page-contents" title="Permalink to this headline">#</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="#GS">Getting Started</a></p></li>
<li><p><a class="reference external" href="#OV">Overview</a></p></li>
<li><p><a class="reference external" href="#SOF">Software Requirements</a></p></li>
</ul>
</section>
<section id="a-name-gs-getting-started-a">
<h2><a name="GS">Getting Started</a><a class="headerlink" href="#a-name-gs-getting-started-a" title="Permalink to this headline">#</a></h2>
<p>Each learning submodules will be organized in a R Jupyter notebook to help the participants familiarize themselves
with the cloud computing in the specific context of running bioinformatics workflows. Each notebook will include
step-by-step hands-on practice with R command line to install necessary tools, obtain data, perform analyses, visualize
and interpret the results.</p>
</section>
<section id="a-name-ov-overview-a">
<h2><a name='OV'>Overview</a><a class="headerlink" href="#a-name-ov-overview-a" title="Permalink to this headline">#</a></h2>
<p>The content of the course is organized in R Jupyter Notebooks. Then we use Jupyter Book which is a package to combine
individuals Jupyter Notebooks into a web-interface for a better navigation. Details of installing the tools and formatting
the content can be found at: <a class="reference external" href="https://jupyterbook.org/en/stable/intro.html">https://jupyterbook.org/en/stable/intro.html</a>. The content of the course is reposed in the
Github repository of Dr. Tin Nguyen’s lab, and can be found at <a class="reference external" href="https://github.com/tinnlab/NOSI-Google-Cloud-Training">https://github.com/tinnlab/NOSI-Google-Cloud-Training</a>.</p>
<p>Module 1 of the course provides basic knowledge of Data Processing. It walks the student through the process of obtaining data from
multiple platforms such as public repositories like the Gene Expression Omnibus (GEO), preprocessing and manipulating this
data, normalization, loading and saving the Expression matrix. It also explains the concept of Probe Set ID and describes
how to map probe IDs into gene identifiers and symbols.</p>
<p>Module 2 describes the next step in the Pathway Analysis workflow after data acquisition and preparation, which is
Differential Expression Analysis. This is an important process, which goal is to determine which genes are expressed at
different levels between two or more biological conditions.</p>
<p>The results of DE analyses are often gene lists, which are difficult to interpret due to their large size and lack of
useful annotations. Hence, Pathway Analysis aims to reduce the complexity of interpreting gene lists via mapping the
listed genes to known biological pathways, processes and functions, which are contained in common curated biological
databases. Module 3 introduces such databases including Gene Ontology (GO), Kyoto Encyclopedia of Genes and Genomes (KEGG).</p>
<p>In Module 4, some Enrichment Analysis methods, which help gain insight into obtained gene lists by identifying pathways that
are enriched in a gene list more than would be expected by chance are introduced. The three major steps involved the process,
including definition of a gene list from omics data, determination of statistically enriched pathways, and visualization
and interpretation of the results are explained.</p>
<p>Given the importance and applications of Pathway Analysis in physiological and biomedical research, many methods
have been developed to identify enriched pathways under different biological states within a genomic study. Also, as more
and more microarray datasets accumulate, meta-analysis methods have also been developed to integrate information among
multiple studies, in a bid to understand the underlying biological phenomena behind the gene expression changes involved
in a comparison of two phenotypes. Module 5 focuses on these methods and explores methods such as Fisher’s and Stouffer’s
methods in detail.</p>
</section>
<section id="software-requirements-a-name-sof-a">
<h2><strong>Software Requirements</strong> <a name="SOF"></a><a class="headerlink" href="#software-requirements-a-name-sof-a" title="Permalink to this headline">#</a></h2>
<p>This learning module does not require any computational hardware and local environment setting from users as the
programs and scripts can be run in the browser-based development environment provided by Google. However,
users need to have Google email account, sufficient internet access, and a standard web-browser (e.g. Chrome, Edge,
Firefox etc., Chrome browser is recommended) to create a cloud virtual machine for analysis.</p>
<p>The Jupyter Book will run and test all codes in our notebooks and then build them into a website format.
The website output is the <code class="docutils literal notranslate"><span class="pre">_build/html</span></code> folder in the repository. After the <code class="docutils literal notranslate"><span class="pre">_build/html</span></code> is completely built, the web
version of the book will be uploaded from the <code class="docutils literal notranslate"><span class="pre">_build/html</span></code> folder to a Google bucket.
The address of the cloud bucket can be found at: <a class="reference external" href="https://storage.googleapis.com/nosi-gcloud-course/html/intro.html">https://storage.googleapis.com/nosi-gcloud-course/html/intro.html</a>.
The link <a class="reference external" href="http://tinnguyen-lab.com/course/cpa-gcloud/intro.html">http://tinnguyen-lab.com/course/cpa-gcloud/intro.html</a> is just a proxy to that link using Dr. Tin Nguyen’s Lab
domain name.</p>
<p>A complete workflow from setting up the environment, building the book, to uploading the website to
Google cloud bucket can be found here: <a class="reference external" href="https://github.com/tinnlab/NOSI-Google-Cloud-Training/blob/main/.github/workflows/main.yml">https://github.com/tinnlab/NOSI-Google-Cloud-Training/blob/main/.github/workflows/main.yml</a>.
The logs for each time it builds the book are here: <a class="reference external" href="https://github.com/tinnlab/NOSI-Google-Cloud-Training/actions">https://github.com/tinnlab/NOSI-Google-Cloud-Training/actions</a>.</p>
</section>
<section id="workflow-diagrams-a-name-work-a">
<h2><strong>Workflow Diagrams</strong> <a name="WORK"></a><a class="headerlink" href="#workflow-diagrams-a-name-work-a" title="Permalink to this headline">#</a></h2>
<blockquote>
<div><p>Explain the cloud architecture/data flow of your workflow here, and include your data flow diagram here.</p>
</div></blockquote>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="img-src-images-intro-main-img-png-width-900-height-700">
<h1><img src="./images/Intro/Main-img.png" width="900" height="700"><a class="headerlink" href="#img-src-images-intro-main-img-png-width-900-height-700" title="Permalink to this headline">#</a></h1>
<p>As seen in the image above, we will show how to download sequence data from sources such as the GEO repository using the
GEO website as well as the GEOquery package from the R programming language terminal.
We will use the getGEO function of the package to download GEO datasets, as identified by the accession ID.
Exploration and preprocessing of the data follows, after which we export it to a desired format.
Next, we map the probe set IDs of the datasets to Entrez gene ID to achieve uniformity, before carrying out Differential
Gene Expression Analysis on the data, which is the focus of Module 2.
It involves assigning samples into groups and setting up design matrix, and then performing DE analysis using limma, t-test,
edgeR and DESeq packages to produce results which are filtered and exported before being further visualized.</p>
<p>Module 3 introduces Ontology, Gene Ontology and the KEGG Pathway Database. Also, terms and pathway gene sets are downloaded
from GO and KEGG and then the results are saved to the GMT file format. In Module 4, some Pathway Analysis methods are introduced, before Meta analysis methods are discussed in Module 5.</p>
<section id="data-a-name-data-a">
<h2><strong>Data</strong> <a name="DATA"></a><a class="headerlink" href="#data-a-name-data-a" title="Permalink to this headline">#</a></h2>
<p>All data from module the was originally downloaded from the Gene Expression Omnibus (GEO) repository using the accession
number GSE48350 file. The data was originally generated by <strong>Berchtold and Cotman, 2013</strong>. We preprocessed this data and normalized it,
after which we used it in the subsequent analyses.</p>
</section>
<section id="troubleshooting-a-name-tr-a">
<h2><strong>Troubleshooting</strong> <a name="TR"></a><a class="headerlink" href="#troubleshooting-a-name-tr-a" title="Permalink to this headline">#</a></h2>
<p>Describe common error encountered with your workflow and how to troubleshoot them
Prior to publishing, some common errors frequently encountered by the early users in the workflow include syntax errors
such as calling a library that has not been initialized.
Also, some users who were still beginners to Bioinformatics made basic errors such as confusing the meanings of terms such as
Pathways and Gene Sets, and misunderstanding what these stand for. However, this is an error that is attributed to limited
knowledge and is cleared by exposure to the first few sections of Module 1.</p>
</section>
<section id="funding-a-name-fund-a">
<h2><strong>Funding</strong> <a name="FUND"></a><a class="headerlink" href="#funding-a-name-fund-a" title="Permalink to this headline">#</a></h2>
<blockquote>
<div><p>Describe where your funding came from. You should list your NIGMS award and any other funding that contributed to building your module.</p>
</div></blockquote>
</section>
<section id="optional-but-recommended-license-for-data-a-name-lic-a">
<h2><strong>(Optional, but recommended) License for Data</strong> <a name="LIC"></a><a class="headerlink" href="#optional-but-recommended-license-for-data-a-name-lic-a" title="Permalink to this headline">#</a></h2>
<p>Text and materials are licensed under a Creative Commons CC-BY-NC-SA license. The license allows you to copy, remix
and redistribute any of our publicly available materials, under the condition that you attribute the work (details in the license)
and do not make profits from it. More information is available <a class="reference external" href="https://tilburgsciencehub.com/about/#license">here</a>.</p>
<p><a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" 
style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br /></p>
<p>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative
Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.</p>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Bang Tran, Hung Nguyen, and Tin Nguyen<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>